rest_api/sawtooth_rest_api/messaging.py

```diff
@@ -0,0 +1,290 @@
+from zmq.asyncio import Poller
+
+from sawtooth_rest_api.protobuf.validator_pb2 import Message
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+class _MessageRouter:
+    """Manages message, routing them either to an incoming queue or to the
+    futures for expected replies.
+    """
+    def __init__(self):
+        self._queue = asyncio.Queue()
+        self._futures = {}
+
+    async def _push_incoming(self, msg):
+        return await self._queue.put(msg)
+
+    async def incoming(self):
+        """Returns the next incoming message.
+        """
+        return await self._queue.get()
+
+    def expect_reply(self, correlation_id):
+        """Informs the router that a reply to the given correlation_id is
+        expected.
+        """
+        self._futures[correlation_id] = asyncio.Future()
+
+    def expected_replies(self):
+        """Returns the correlation ids for the expected replies.
+        """
+        return (c_id for c_id in self._futures)
+
+    async def await_reply(self, correlation_id, timeout=None):
+        """Wait for a reply to a given correlation id.  If a timeout is
+        provided, it will raise a asyncio.TimeoutError.
+        """
+        result = await asyncio.wait_for(
+            self._futures[correlation_id], timeout=timeout)
+
+        del self._futures[correlation_id]
+
+        return result
+
+    def _set_reply(self, correlation_id, msg):
+        if correlation_id in self._futures:
+            self._futures[correlation_id].set_result(msg)
+
+    def _fail_reply(self, correlation_id, err):
+        if correlation_id in self._futures and \
+                not self._futures[correlation_id].done():
+            self._futures[correlation_id].set_exception(err)
+
+    def fail_all(self, err):
+        """Fail all the expected replies with a given error.
+        """
+        for c_id in self._futures:
+            self._fail_reply(c_id, err)
+
+    async def route_msg(self, msg):
+        """Given a message, route it either to the incoming queue, or to the
+        future associated with its correlation_id.
+        """
+        if msg.correlation_id in self._futures:
+            self._set_reply(msg.correlation_id, msg)
+        else:
+            await self._push_incoming(msg)
+
+
+class _Receiver:
+    """Receives messages and forwards them to a _MessageRouter.
+    """
+
+    def __init__(self, socket, msg_router):
+        self._socket = socket
+        self._msg_router = msg_router
+
+        self._is_running = False
+
+    async def start(self):
+        """Starts receiving messages on the underlying socket and passes them
+        to the message router.
+        """
+        poller = Poller()
+        poller.register(self._socket, zmq.POLLIN)
+        self._is_running = True
+
+        while self._is_running:
+            try:
+                events = await poller.poll()
+                if self._socket in dict(events):
+                    zmq_msg = await self._socket.recv_multipart()
```

Comment:
Since this is a Dealer socket and we only want the bytes transported, I think self._socket.recv() will work.